moler/util/moler_test.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+"""
+Utility/common code of library.
+"""
+
+__author__ = 'Grzegorz Latuszek, Marcin Usielski, Michal Ernst'
+__copyright__ = 'Copyright (C) 2018, Nokia'
+__email__ = 'grzegorz.latuszek@nokia.com, marcin.usielski@nokia.com, michal.ernst@nokia.com'
+
+import logging
+import time
+from functools import wraps
+from types import FunctionType
+
+from moler.connection_observer import ConnectionObserver
+from moler.exceptions import MolerException
+
+
+class MolerTest(object):
+    _last_check_time = 0
+    _was_error = False
+    _was_steps_end = False
+    _logger = logging.getLogger("moler")
+
+    @staticmethod
+    def steps_end():
+        MolerTest._was_steps_end = True
+
+    @staticmethod
+    def log_error(msg, raise_exception=False):
+        MolerTest._was_error = True
+        MolerTest._logger.error(msg)
+
+        if raise_exception:
+            raise MolerException(msg)
+
+    @staticmethod
+    def log(msg):
+        MolerTest._logger.info(msg)
+
+    @staticmethod
+    def _steps_start():
+        MolerTest._was_steps_end = False
+
+    @staticmethod
+    def _final_check():
+        # Checks exceptions since last call final_check
+        final_check_time = time.time()
+        exceptions = ConnectionObserver.get_active_exceptions_in_time(MolerTest._last_check_time)
+        for exception in exceptions:
+            MolerTest.log_error("Unhandled exception: '{}'".format(exception))
+        MolerTest._last_check_time = final_check_time
+        was_error_in_last_execution = MolerTest._was_error
+        MolerTest._was_error = False
+        assert MolerTest._was_steps_end is True
+        assert was_error_in_last_execution is False
+
+    @staticmethod
+    def moler_test_status():
+        def decorate(cls):
+            for attributeName, attribute in cls.__dict__.items():
+                if attributeName.startswith("test"):
+                    if isinstance(attribute, FunctionType):
+                        setattr(cls, attributeName, MolerTest.wrapper(attribute))
+            return cls
+
+        return decorate
+
+    @staticmethod
```

Comment:
no need for `@staticmethod` - put it outside class

Reply:
No need however from my perspective it is more readable inside class.

Reply:
I agree with Marcin

Reply:
Disagree to leave `wrapper()` and `moler_test_status()` inside class as `@staticmethod`. I would put them as pure functions and renamed and combined:

- renamed to `reraise_background_exceptions()` since such name explains what happens inside decorator while `moler_test_status()` says nothing besides convention "you should use it"
- combining both is possible -  `reraise_background_exceptions()` at first step would check what it is decorating. If class then we call different code, if function/method then different. Thanks to that we can use same decorator on pure-function-keywords and on class-implementing-keywords. **It simplifies life of end user**.

Reply:
And I disagree to give them outside class:) As static method they can use protected/private members. Outside class we can still easily get access to these members but this looks ugly. We can change them to public ones but this way we don't show user not to use them. Python allows to make a mess:).

Reply:
I think everything is discussed during f2f review.